packages/artifact/src/internal/download-http-client.ts

```diff
@@ -0,0 +1,195 @@
+   */
+  async listArtifacts(): Promise<ListArtifactsResponse> {
+    const artifactUrl = getArtifactUrl()
+    // use the first client from the httpManager, `keep-alive` is not used so the connection will close immediatly
+    const client = this.downloadHttpManager.getClient(0)
+    const requestOptions = getRequestOptions('application/json')
+
+    const rawResponse = await client.get(artifactUrl, requestOptions)
+    const body: string = await rawResponse.readBody()
+    if (isSuccessStatusCode(rawResponse.message.statusCode) && body) {
+      return JSON.parse(body)
+    }
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(`Unable to list artifacts for the run`)
+  }
+
+  /**
+   * Fetches a set of container items that describe the contents of an artifact
+   * @param artifactName the name of the artifact
+   * @param containerUrl the artifact container URL for the run
+   */
+  async getContainerItems(
+    artifactName: string,
+    containerUrl: string
+  ): Promise<QueryArtifactResponse> {
+    // the itemPath search parameter controls which containers will be returned
+    const resourceUrl = new URL(containerUrl)
+    resourceUrl.searchParams.append('itemPath', artifactName)
+
+    // no concurrent calls so a single httpClient without the http-manager is sufficient
+    const client = createHttpClient()
+
+    // no keep-alive header, client disposal is not necessary
+    const requestOptions = getRequestOptions('application/json')
+    const rawResponse = await client.get(resourceUrl.toString(), requestOptions)
+    const body: string = await rawResponse.readBody()
+    if (isSuccessStatusCode(rawResponse.message.statusCode) && body) {
+      return JSON.parse(body)
+    }
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(`Unable to get ContainersItems from ${resourceUrl}`)
+  }
+
+  /**
+   * Concurrently downloads all the files that are part of an artifact
+   * @param downloadItems information about what items to download and where to save them
+   */
+  async downloadSingleArtifact(downloadItems: DownloadItem[]): Promise<void> {
+    const DOWNLOAD_CONCURRENCY = getDownloadFileConcurrency()
+    // limit the number of files downloaded at a single time
+    const parallelDownloads = [...new Array(DOWNLOAD_CONCURRENCY).keys()]
+    let downloadedFiles = 0
+    await Promise.all(
+      parallelDownloads.map(async index => {
+        while (downloadedFiles < downloadItems.length) {
+          const currentFileToDownload = downloadItems[downloadedFiles]
+          downloadedFiles += 1
+          await this.downloadIndividualFile(
+            index,
+            currentFileToDownload.sourceLocation,
+            currentFileToDownload.targetPath
+          )
+        }
+      })
+    )
+
+    // done downloading, safety dispose all connections
+    this.downloadHttpManager.disposeAndReplaceAllClients()
```

Comment:
You probably should separate those two functions - here there is no need to replace all clients.

Reply:
Disposal and replacement is necessary when calling `downloadAllArtifacts` because the same `downloadHttpClient` gets used sequentially and we need to guarantee that the clients are properly replaced and disposed.

When uploading an artifact or downloading an individual one, the extra replacement is a little wasteful because some of the clients will never be used again, but it's safer to properly dispose and replace each client so that it's guaranteed that all clients can be used if need be